packages/graphile-build-pg/src/QueryBuilder.js

```diff
@@ -495,38 +501,51 @@ class QueryBuilder {
     } else if (type === "whereBound") {
       // Handle properties separately
       this.compiledData[type].lower = callIfNecessaryArray(
-        this.data[type].lower
+        this.data[type].lower,
+        { queryBuilder: this }
       );
       this.compiledData[type].upper = callIfNecessaryArray(
-        this.data[type].upper
+        this.data[type].upper,
+        { queryBuilder: this }
       );
     } else if (type === "select") {
       this.compiledData[type] = this.data[type].map(([a, b]) => [
-        callIfNecessary(a),
+        callIfNecessary(a, { queryBuilder: this }),
         b,
       ]);
     } else if (type === "orderBy") {
       this.compiledData[type] = this.data[type].map(([a, b]) => [
-        callIfNecessary(a),
+        callIfNecessary(a, { queryBuilder: this }),
         b,
       ]);
     } else if (type === "from") {
       if (this.data.from) {
         const f = this.data.from;
-        this.compiledData.from = [callIfNecessary(f[0]), f[1]];
+        this.compiledData.from = [
+          callIfNecessary(f[0], { queryBuilder: this }),
+          f[1],
+        ];
       }
     } else if (type === "join" || type === "where") {
-      this.compiledData[type] = callIfNecessaryArray(this.data[type]);
+      this.compiledData[type] = callIfNecessaryArray(this.data[type], {
+        queryBuilder: this,
+      });
     } else if (type === "selectCursor") {
-      this.compiledData[type] = callIfNecessary(this.data[type]);
+      this.compiledData[type] = callIfNecessary(this.data[type], {
+        queryBuilder: this,
+      });
     } else if (type === "cursorPrefix") {
       this.compiledData[type] = this.data[type];
     } else if (type === "orderIsUnique") {
       this.compiledData[type] = this.data[type];
     } else if (type === "limit") {
-      this.compiledData[type] = callIfNecessary(this.data[type]);
+      this.compiledData[type] = callIfNecessary(this.data[type], {
+        queryBuilder: this,
+      });
     } else if (type === "offset") {
-      this.compiledData[type] = callIfNecessary(this.data[type]);
+      this.compiledData[type] = callIfNecessary(this.data[type], {
+        queryBuilder: this,
+      });
```

Comment:
Let's DRY this up by putting something like:

```js
const getContext = () => ({
  queryBuilder: this,
})
```
at the top of the `lock` method and then calling it at each place instead:

```js
this.compiledData[type] = callIfNecessary(this.data[type], getContext());
```